consensus/src/validation/step.rs

```diff
@@ -54,20 +54,41 @@ impl<T: Operations + 'static> ValidationStep<T> {
         inbound: AsyncQueue<Message>,
         executor: Arc<Mutex<T>>,
     ) {
-        // TODO: Verify Block Header
         let hash = candidate.header().hash;
 
-        // Call VST for non-empty blocks
-        if hash != [0u8; 32] {
-            if let Err(err) = Self::call_vst(candidate, ru, executor).await {
-                error!(
-                    event = "failed_vst_call",
-                    reason = format!("{:?}", err)
-                );
-                return;
-            }
+        // A Validation step with empty/default Block produces a Nil Vote
+        if hash == [0u8; 32] {
+            Self::cast_vote([0u8; 32], ru, iteration, outbound, inbound).await;
+            return;
+        }
+
+        // Verify candidate header all fields except the winning certificate
+        // NB: Winning certificate is produced only on reaching consensus
+        if let Err(err) = executor
+            .lock()
+            .await
+            .verify_block_header(candidate.header(), true)
+            .await
+        {
+            error!(event = "invalid_header", ?err);
+            return;
+        };
+
+        // Call Verify State Transition to make sure transactions set is valid
+        if let Err(err) = Self::call_vst(candidate, ru, executor).await {
+            error!(event = "failed_vst_call", ?err);
```

Comment:
Should this `return`?

Reply:
Then you should add the "return" otherwise cast_vote is called

Reply:
Good catch!